### PR TITLE
Proxy Handler Setup 

### PR DESCRIPTION
**Pull request by binerys**
Hi! Here's the initial code layout for the proxy handler. Let me know if there are things I should change, such as the naming conventions used in the config for proxy handling. Furthermore, I'm not sure what convention you follow if there are missing config values (ie they fail to specify the proxy host) so I'm sticking with the NOT_FOUND status for now so the NotFoundHandler will deal with it.

### CODE REVIEW

File: request_handlers.cc [hunk 1]
```diff
@@ -156,6 +156,40 @@ RequestHandler::Status NotFoundRequestHandler::handleRequest(const Request& requ
   return RequestHandler::NOT_FOUND;
 }
 
+bool ProxyRequestHandler::init(const std::string& uri_prefix,
+                                  const NginxConfig& config) {
```
Comment (KyCodeHuynh) [inline request_handlers.cc]: It looks like this second param got misaligned?

File: request_handlers.cc [hunk 2]
```diff
@@ -156,6 +156,40 @@ RequestHandler::Status NotFoundRequestHandler::handleRequest(const Request& requ
...
+  uri_prefix_ = uri_prefix;
+  config_ = ServerConfig(config);
+  // TODO: do logging/check return value of build
+  config_.build();
+  return true;
+}
+
+RequestHandler::Status ProxyRequestHandler::handleRequest(const Request& request,
+                                                             Response* response) {
```
Comment (KyCodeHuynh) [inline request_handlers.cc]: This second param is also misaligned vertically.

File: request_handlers.cc [hunk 3]
```diff
@@ -156,6 +156,40 @@ RequestHandler::Status NotFoundRequestHandler::handleRequest(const Request& requ
...
+  std::cout << "ProxyRequestHandler currently responding.\n";
+
+  std::string proxy_host;
+  std::string proxy_port; 
+  std::vector<std::string> query_host = {"proxy_host"};
+  std::vector<std::string> query_port = {"proxy_port"};
+
+  config_.printPropertiesMap();
+  if(!config_.propertyLookUp(query_host, proxy_host)) {
```
Comment (KyCodeHuynh) [inline request_handlers.cc]: Our convention for if-blocks is: 

```c++
if (condition) {
    // Code
}
```

We just need a space after the `if`. 👍

File: request_handlers.cc [hunk 4]
```diff
@@ -156,6 +156,40 @@ RequestHandler::Status NotFoundRequestHandler::handleRequest(const Request& requ
...
+    std::cout << "Failed to specify proxy host.\n";
+    return RequestHandler::NOT_FOUND; 
+  }
+
+  if(!config_.propertyLookUp(query_port, proxy_port)) {
```
Comment (KyCodeHuynh) [inline request_handlers.cc]: [Same as above](https://github.com/UCLA-CS130/Mr.-Robot-et-al./pull/64/files#r103375659)

**Comment by KyCodeHuynh** [comment]
You're totally right that we're using `RequestHandler::NOT_FOUND` for now, at least until we introduce a `SERVER_ERROR` (5xx errors) later on.

File: request_handlers.h [hunk 5]
```diff
@@ -109,6 +109,22 @@ class NotFoundRequestHandler : public RequestHandler {
 };
 REGISTER_REQUEST_HANDLER(NotFoundRequestHandler);
 
+// Handle proxy server requests
+class ProxyRequestHandler : public RequestHandler {
```
Comment (KyCodeHuynh) [inline request_handlers.h]: Could we call this `ReverseProxyRequestHandler` instead? That way, we can differentiate this vs. a potential bonus feature of (forward) proxy functionality. 😸

File: request_handlers.h [hunk 6]
```diff
@@ -109,6 +109,22 @@ class NotFoundRequestHandler : public RequestHandler {
...
+  public:
+    ProxyRequestHandler() {}
+
+    virtual bool init(const std::string& uri_prefix,
+                        const NginxConfig& config);
```
Comment (KyCodeHuynh) [inline request_handlers.h]: This param was also misaligned. Perhaps this is a setting in your editor?

File: request_handlers.h [hunk 7]
```diff
@@ -109,6 +109,22 @@ class NotFoundRequestHandler : public RequestHandler {
...
+
+    virtual Status handleRequest(const Request& request,
+                                 Response* response);
+  private:
+    std::string uri_prefix_;
```
Comment (KyCodeHuynh) [inline request_handlers.h]: Nice job following the variable-naming conventions!

File: simple_config [hunk 8]
```diff
@@ -12,5 +12,10 @@ path /echo EchoRequestHandler {}
 
 path /status StatusHandler {}
 
+path /proxy ProxyRequestHandler {
```
Comment (KyCodeHuynh) [inline simple_config]: On a similar note to the above, could we name this `/reverse_proxy ReverseProxyRequestHandler`?

File: simple_config [hunk 9]
```diff
@@ -12,5 +12,10 @@ path /echo EchoRequestHandler {}
...
+  proxy_host  ucla.edu;
```
Comment (KyCodeHuynh) [inline simple_config]: And finally, just for consistency, we can call this `reverse_proxy_host`, which would prevent us accidentally mixing up host/port settings between proxy and reverse-proxy config-blocks, for example.

**Comment by binerys** [comment]
@KyCodeHuynh updated the branch with the changes! Thanks for the review 👍

**Pull request merged**